ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml

```diff
@@ -0,0 +1,1911 @@
+external set_scannable :
+  #(int * float * string) array -> nativeint# -> #(int * float * string)
+  -> unit =
+  "%array_unsafe_set_indexed_by_nativeint#"
+let set_scannable_app a i x = set_scannable a i x
+
+external set_ignorable :
+  #(float# * int * int64# * bool) array -> nativeint#
+  -> #(float# * int * int64# * bool) -> unit =
+  "%array_unsafe_set_indexed_by_nativeint#"
+let set_ignorable_app a i x = set_ignorable a i x
+[%%expect{|
+external set_scannable :
+  #(int * float * string) array ->
+  nativeint# -> #(int * float * string) -> unit
+  = "%array_unsafe_set_indexed_by_nativeint#"
+val set_scannable_app :
+  #(int * float * string) array ->
+  nativeint# -> #(int * float * string) -> unit = <fun>
+external set_ignorable :
+  #(float# * int * int64# * bool) array ->
+  nativeint# -> #(float# * int * int64# * bool) -> unit
+  = "%array_unsafe_set_indexed_by_nativeint#"
+val set_ignorable_app :
+  #(float# * int * int64# * bool) array ->
+  nativeint# -> #(float# * int * int64# * bool) -> unit = <fun>
+|}]
+
+external set_bad :
+  #(string * float#) array -> nativeint# -> #(string * float#) -> unit =
+  "%array_unsafe_set_indexed_by_nativeint#"
+let set_bad_app a i x = set_bad a i x
+[%%expect{|
+external set_bad :
+  #(string * float#) array -> nativeint# -> #(string * float#) -> unit
+  = "%array_unsafe_set_indexed_by_nativeint#"
+Line 4, characters 24-37:
+4 | let set_bad_app a i x = set_bad a i x
+                            ^^^^^^^^^^^^^
+Error: Unboxed product array elements must be external or contain all gc
+       scannable types. The product type this function is applied at is
+       not external but contains an element of sort float64.
+|}]
+
+(* Unboxed vectors are also rejected. *)
+let f_bad (x : #(int * int32x4#) array) v = set x #42n #(1, v)
+[%%expect{|
+Line 1, characters 44-62:
+1 | let f_bad (x : #(int * int32x4#) array) v = set x #42n #(1, v)
+                                                ^^^^^^^^^^^^^^^^^^
+Error: Unboxed vector types are not yet supported in arrays of unboxed
+       products.
+|}]
+
+external set_ignorable_with_vec :
+  #(int * int32x4#) array -> nativeint# -> #(int * int32x4#) -> unit =
+  "%array_unsafe_set_indexed_by_nativeint#"
+let set_ignorable_with_vec_app x i v = set_ignorable_with_vec x i #(1, v)
+[%%expect{|
+external set_ignorable_with_vec :
+  #(int * int32x4#) array -> nativeint# -> #(int * int32x4#) -> unit
+  = "%array_unsafe_set_indexed_by_nativeint#"
+Line 4, characters 39-73:
+4 | let set_ignorable_with_vec_app x i v = set_ignorable_with_vec x i #(1, v)
+                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: Unboxed vector types are not yet supported in arrays of unboxed
+       products.
+|}]
+
+(**********************)
+(* Test 22: arrayblit *)
+
+(* An array poly version works at valid product layouts. *)
+external[@layout_poly] blit :
+  ('a : any_non_null) . 'a array -> int -> 'a array -> int -> int -> unit =
```

Comment:
I agree they are going to want such things, but think we should leave it for a future PR as it will involve either substantially generalizing or duplicating the two new magic externals (`%makearray_dynamic` and `%arrayblit`) and their implementations in the middle and backend are going to be complicated enough already.

(Also I think 'returned to the manufacturer' probably overstates it: they will be annoyed at having to add some extra `Int64_u.to_int`s, but it's not a fundamental obstacle)

Reply:
(I've made a ticket for this)

Reply:
I guess maybe I'm wondering if we need the tagged primitives at all. Maybe we just offer the untagged ones, and implement any desired tagged version in terms of an untagged primitive.

Reply:
I think the primitives for tagged ints are essential.  `int` remains a pervasive and convenient to use type, and it's totally reasonable to use it with these arrays - saving allocations and indirections is more important than saving a bitshift in an indexing operation.  It's clear we're going to need all the others too, and will add them in due course.

Reply:
I'm fine with leaving this Resolved, but it seems simpler to me to implement a primitive over untagged things and then just have a little wrapper around it for tagged ones. That is, I would want the tagged version available to users, as you suggest, but I don't see why that tagged version needs to be a primitive. Maybe it would be more efficient? But I would hope not, with our optimizer.

Reply:
Yes, sorry, I misread your comment.  I agree that potentially makes sense, and we could consider switching to it in a future PR that adds the untagged versions.